canals/component/input_output.py

```diff
@@ -2,78 +2,151 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 import logging
-import inspect
-from dataclasses import fields
+from enum import Enum
+from dataclasses import fields, is_dataclass, dataclass, asdict, MISSING
+
+from canals.errors import ComponentError
 
 logger = logging.getLogger(__name__)
 
 
-class BaseIODataclass:  # pylint: disable=too-few-public-methods
+def _make_fields_optional(class_: type):
     """
-    Base class for input and output classes of components.
+    Takes a dataclass definition and modifies its __init__ so that all fields have
+    a default value set.
+    If a field has a default factory use it to set the default value.
+    If a field has neither a default factory or value default to None.
     """
+    defaults = []
+    for field in fields(class_):
+        default = field.default
+        if field.default is MISSING and field.default_factory is MISSING:
+            default = None
+        elif field.default is MISSING and field.default_factory is not MISSING:
+            default = field.default_factory()
```

Comment:
That won't happen, `default_factory` is used exactly to avoid that trap.

If you try to create this dataclass as an example it would fail.

```
@dataclass
class Foo:
     a: List = field(default_factory=list)
     b: List = field(default=[])

ValueError: mutable default <class 'list'> for field b is not allowed: use default_factory
```

Using a factory is safe:

```
In [60]: @dataclass
    ...: class Foo:
    ...:     a: List = field(default_factory=list)
    ...:     b: int = field(default=1)
    ...:

In [61]: Foo.__init__.__defaults__
Out[61]: (<factory>, 1)

In [62]: Foo()
Out[62]: Foo(a=[], b=1)

In [63]: Foo()
Out[63]: Foo(a=[], b=1)

In [64]: Foo()
Out[64]: Foo(a=[], b=1)

In [65]: Foo()
Out[65]: Foo(a=[], b=1)
```

I would support the defaults of dataclasses, would be confusing if they're not used. We might remove defaults some other places if there are too many places where they can be set.

Reply:
Oh ok, I didn't know dataclasses fail the default factory if you use an empty list. Good then!

> I would support the defaults of dataclasses, would be confusing if they're not used. We should remove defaults some other places if there are too many places where they can be set.

I'd rather fail the object initialization if there are defaults. The other place is the component's `__init__` which I find uglier to mess with. Not for this PR anyway, though. I'll open a separate issue to remember about this,

Reply:
https://github.com/deepset-ai/canals/issues/24